### PR TITLE
Notification center

### PR DESCRIPTION
**Pull request by AbdurRafay**
* Listeners added
* Send notifications
* Test cases
* Test case for sendNotifications
* refactoring
* copyrights changed
* redesigned

### CODE REVIEW

File: OptimizelySDKCore/OptimizelySDKCore/Optimizely.m [hunk 1]
```diff
@@ -354,9 +352,10 @@ - (void)track:(NSString *)eventKey
...
         userInfo[OptimizelyNotificationsUserDictionaryExperimentVariationMappingKey] = [experimentVariationMapping copy];
     }
-    [[NSNotificationCenter defaultCenter] postNotificationName:OptimizelyDidTrackEventNotification
-                                                        object:self
-                                                      userInfo:userInfo];
+    NSString *_userId = userId ? userId : @"";
+    NSDictionary *_attributes = attributes ? attributes : [NSDictionary new];
+    NSDictionary *_eventTags = eventTags ? eventTags : [NSDictionary new];
+    [_notificationCenter sendNotifications:OPTLYNotificationTypeTrack args:eventKey, _userId, _attributes, _eventTags, conversionEventParams, nil];
 }
```
Comment (AbdurRafay) [inline OptimizelySDKCore/OptimizelySDKCore/Optimizely.m]: In order to implement variable arguments in sendNotifications method, I make sure no argument could be pass as nil which can break the argument list.

File: OptimizelySDKCore/OptimizelySDKCore/Optimizely.m [hunk 2]
```diff
@@ -423,6 +422,9 @@ - (OPTLYVariation *)sendImpressionEventFor:(OPTLYExperiment *)experiment
                                                  callback(error);
                                              }
                                          }];
+    NSString *_userId = userId ? userId : @"";
+    NSDictionary *_attributes = attributes ? attributes : [NSDictionary new];
+    [_notificationCenter sendNotifications:OPTLYNotificationTypeActivate args:experiment, _userId, _attributes, variation, impressionEventParams, nil];
     return variation;
```
Comment (AbdurRafay) [inline OptimizelySDKCore/OptimizelySDKCore/Optimizely.m]: Same here.
In order to implement variable arguments in sendNotifications method, I make sure no argument could be pass as nil which can break the argument list.

**Comment by msohailhussain** [comment]
@thomaszurkan-optimizely @kellyroach-optimizely Can you please review this PR.

**Comment by kellyroach-optimizely** [comment]
This is for OASIS-2080 .

**Review by thomaszurkan-optimizely — CHANGES_REQUESTED** [review]
This PR looks good for the Notificaiton Listeners part.  But, there are feature variable checkins.  Can we separate that?

**Review by thomaszurkan-optimizely — APPROVED** [review]
Never mind, I see that it is just formatting and no code change.  So, I'm approving!  thanks!

**Comment by kellyroach-optimizely** [comment]
ISSUE: Use OPTLYLogger and programmed error recovery instead of "assert"

The P.R. #210 would add many "assert"s to OPTLYNotificationCenter.m but this
would be precedent setting for the SDK .  It could be discussed, but I think I would
not like to see this happen.

Instead, use OPTLYLogger for logging failed a modest but appropriate number
of input sanity checks like most of the rest of the iOS SDK code does, and arrange
that callers of your code do recover gracefully (do not crash) if OPTLYNotificationCenter
methods reject some occasional incorrect inputs.  Large complex software in the
modern era should do this because glitches are inevitable.

OPTLYNotificationCenter does have a
@property (nonatomic, strong) id<OPTLYLogger> logger
The constructor for OPTLYNotificationCenter ought to have a signature like
- (instancetype) initWithProjectConfig:(OPTLYProjectConfig *)config
Then the OPTLYNotificationCenter instance methods themselves should use the config.logger .

OPTLYNotificationCenter should not create its own logger but use the logger
provided by the OPTLYProjectConfig that Optimizely was initialized with.

Therefore, I would like to ask that "assert" not be used in checked in code.
Instead, use OPTLYLogger and programmed error recovery for diagnostics
and avoidance of iOS SDK crashing apps in the field.

**Comment by kellyroach-optimizely** [comment]
ISSUE: Use NSArray inputs instead of va_list va_start va_end

Function sendNotifications:args: is using variable-length argument lists with signature
``- (void)sendNotifications:(OPTLYNotificationType)type args:(id)firstArg, ...``
and relying on using va_list va_start va_end in its body.

For the most part, the iOS SDK avoids va_list va_start va_end, and I think that is a
good practice.  There are a number of reasons.  One reason is that "..." doesn't
allow functional composition.  IOW, the output of one function which could be an
array, can't be used as the input args to sendNotifications:args: as it is currently coded.

**Review by kellyroach-optimizely — CHANGES_REQUESTED** [review]
I have 3 main issues with P.R. #210 .
ISSUE: Use OPTLYLogger and programmed error recovery instead of "assert"
ISSUE: Use NSArray inputs instead of va_list va_start va_end
ISSUE: Use OOP object delegates instead of function pointers

**Comment by kellyroach-optimizely** [comment]
ISSUE: Use OOP object delegates instead of function pointers

OPTLYActivateNotificationListener and OPTLYTrackNotificationListener in
OPTLYNotificationCenter.h are C function pointer types and are being used
in the method signatures
``- (NSInteger)addNotification:(OPTLYNotificationType)type activateListener:(OPTLYActivateNotificationListener)activateListener``
``- (NSInteger)addNotification:(OPTLYNotificationType)type trackListener:(OPTLYTrackNotificationListener)trackListener``
in OPTLYNotificationCenter.m .

I think this raises some issues for OOP style and Swift translation.
I would like to suggest instead creating an Objective-C @protocol with
two method declarations essentially equivalent to the two C function pointer types
you are trying to introduce, and then redeclare the two OPTLYNotificationCenter.m
method signatures to input arguments conforming to this Objective-C @protocol .
E.G. if we name the protocol "OPTLYNotificationDelegate" then the two method
signatures can input ``id<OPTLYNotificationDelegate> activateListener`` and
``id<OPTLYNotificationDelegate> trackListener`` instead.

File: OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYNotificationCenterTest.m [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    // Verify that only decision callbacks are removed.
+    [_notificationCenter clearNotifications:OPTLYNotificationTypeActivate];
+    XCTAssertEqual(1, _notificationCenter.notificationsCount);
+    
+    // Verify that ClearNotifications does not break on calling twice for same type.
+    [_notificationCenter clearNotifications:OPTLYNotificationTypeActivate];
+    [_notificationCenter clearNotifications:OPTLYNotificationTypeActivate];
+    
+    // Verify that ClearNotifications does not break after calling ClearAllNotifications.
+    [_notificationCenter clearAllNotifications];
```
Comment (kellyroach-optimizely) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYNotificationCenterTest.m]: Some code comments mention "ClearAllNotifications", but should say "clearAllNotifications"
with lower case "c" .

**Comment by kellyroach-optimizely** [comment]
CODE REVIEW 180126 AFTER https://github.com/optimizely/objective-c-sdk/pull/210/commits/cfbfe24ff5d44f3eceb80a773bdf1c4c347e6e25

The updated P.R. #210 is very much improved.  Thank you!

I did spot a couple little more items that could be fixed.  These items pertain mainly
to the OPTLYActivateNotificationTest and public notificationsCount .  See comments
about these below.

**Comment by kellyroach-optimizely** [comment]
OPTLYActivateNotificationTest COMPILER WARNINGS
```
.../OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYNotificationCenterTest.m:36:17: warning: class 'OPTLYActivateNotificationTest' does not conform to protocol 'OPTLYNotificationDelegate' [-Wprotocol]
@implementation OPTLYActivateNotificationTest
                ^
.../OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYNotificationCenterTest.m:36:17: note: add stubs for missing protocol requirements
@implementation OPTLYActivateNotificationTest
                ^
.../OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYNotificationCenterTest.m:44:17: warning: class 'OPTLYTrackNotificationTest' does not conform to protocol 'OPTLYNotificationDelegate' [-Wprotocol]
@implementation OPTLYTrackNotificationTest
                ^
.../OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYNotificationCenterTest.m:44:17: note: add stubs for missing protocol requirements
@implementation OPTLYTrackNotificationTest
                ^
2 warnings generated.
```

**Comment by kellyroach-optimizely** [comment]
COMMENT ABOUT PUBLIC "notificationsCount"
```
@property (nonatomic, readonly) NSUInteger notificationsCount;
```
This property is publicly exposed in OPTLYNotificationCenter.h .   However,
our Notification Center design doc does not ask for this, sibling reference python-sdk
does not implement this, and I also asked Thomas Zurkan who wrote the design doc
about "notificationsCount" .  Tom Zurkan didn't think "notificationsCount" is necessary.

Since I do see the "notificationsCount" is being used by the OPTLYActivateNotificationTest
which is only a unit test and not part of our shipped SDK *.framework, I'm wondering and
would like to suggest to you that "notificationsCount" could be moved out of
OPTLYNotificationCenter.m+.h and into OPTLYActivateNotificationTest.m instead as
member of a private OPTLYNotificationCenter extension category visible only to
OPTLYActivateNotificationTest.m .  Could you consider doing this?

**Comment by AbdurRafay** [comment]
I moved out "notificationsCount" from OPTLYNotificationCenter.h and made it a private member of OPTLYNotificationCenter extension in OPTLYNotificationCenterTest.m

Since its implementation in OPTLYNotificationCenter.m also used other private members of OPTLYNotificationCenter like ("notifications") which then should also be exposed in extension.

**Comment by kellyroach-optimizely** [comment]
TEST:
* sh .Scripts/build_all.sh PASSED
* sh .Scripts/local_travis.sh PASSED
* OPTLYNotificationCenterTest.m zero warnings PASSED
* 4 Analyze PASSED
* 8+1 iOS test suites PASSED
* 7+1 tvOS test suites PASSED
* 2 DemoApp's PASSED

All code review comments addressed.  APPROVING.

**Review by kellyroach-optimizely — APPROVED** [review]

**Pull request merged**